src/algebra/quandle.lean

```diff
@@ -0,0 +1,585 @@
+  self_distrib := λ x y z, begin
+    dsimp only [mul_equiv.to_equiv_apply, mul_aut.conj_apply, conj],
+    group,
+  end,
+  inv_act := (λ x, (@mul_aut.conj G _ x).symm),
+  left_inv := λ x y, by { dsimp [act, conj], group },
+  right_inv := λ x y, by { dsimp [act, conj], group },
+  fix := λ x, by simp }
+
+@[simp]
+lemma conj_act_eq_conj {G : Type*} [group G] (x y : conj G) :
+  x ◃ y = ((x : G) * (y : G) * (x : G)⁻¹ : G) := rfl
+
+lemma conj_swap {G : Type*} [group G] (x y : conj G) :
+  x ◃ y = y ↔ y ◃ x = x :=
+begin
+  dsimp, split,
+  repeat { intro h, conv_rhs { rw eq_mul_inv_of_mul_eq (eq_mul_inv_of_mul_eq h) }, simp, },
+end
+
+/--
+`conj` is functorial
+-/
+def conj.map {G : Type*} {H : Type*} [group G] [group H] (f : G →* H) : conj G →◃ conj H :=
+{ to_fun := f,
+  map_act' := by simp }
+
+instance {G : Type*} {H : Type*} [group G] [group H] : has_lift (G →* H) (conj G →◃ conj H) :=
+{ lift := conj.map }
+
+/--
+The dihedral quandle. This is the conjugation quandle of the dihedral group restrict to flips.
+
+Used for Fox n-colorings of knots.
+-/
+@[nolint has_inhabited_instance]
+def dihedral (n : ℕ) := zmod n
+
+/--
+The operation for the dihedral quandle.  It does not need to be an equivalence
+because it is an involution (see `dihedral_act.inv`).
+-/
+def dihedral_act (n : ℕ) (a : zmod n) : zmod n → zmod n :=
+λ b, 2 * a - b
+
+lemma dihedral_act.inv (n : ℕ) (a : zmod n) : function.involutive (dihedral_act n a) :=
+by { intro b, dsimp [dihedral_act], ring }
+
+instance (n : ℕ) : quandle (dihedral n) :=
+{ act := dihedral_act n,
+  self_distrib := λ x y z, begin
+    dsimp [function.involutive.to_equiv, dihedral_act], ring,
+  end,
+  inv_act := dihedral_act n,
+  left_inv := λ x, (dihedral_act.inv n x).left_inverse,
+  right_inv := λ x, (dihedral_act.inv n x).right_inverse,
+  fix := λ x, begin
+    dsimp [function.involutive.to_equiv, dihedral_act], ring,
+  end }
+
+end quandle
+
+namespace rack
+
+/--
+This is the natural rack homomorphism to the conjugation quandle of the group `R ≃ R`
+that acts on the rack.
+-/
+def to_conj (R : Type*) [rack R] : R →◃ quandle.conj (R ≃ R) :=
+{ to_fun := act,
+  map_act' := ad_conj }
+
+section envel_group
+
+/-!
+## Universal enveloping group of a rack
+
+Joyce, for quandles, called this AdConj.
+-/
+
+/--
+Free generators of the enveloping group.
+-/
+inductive pre_envel_group (R : Type u) : Type u
+| unit : pre_envel_group
+| incl (x : R) : pre_envel_group
+| mul (a b : pre_envel_group) : pre_envel_group
+| inv (a : pre_envel_group) : pre_envel_group
+
+instance pre_envel_group.inhabited (R : Type u) : inhabited (pre_envel_group R) :=
+⟨pre_envel_group.unit⟩
+
+open pre_envel_group
+
+/--
+Relations for the enveloping group. This is a type-valued relation because
+`to_envel_group.map_aux.well_def` inducts on it to show `to_envel_group.map`
+is well-defined.  The relation `pre_envel_group_rel` is the `Prop`-valued version,
+which is used to define `envel_group` itself.
+-/
```

Comment:
I still don't understand what's going on here. Why can't `to_envel_group.map_aux.well_def` cope if we just make this `Prop` valued right away?

Reply:
I can't say whether or not there's another way to organize things so that the relation is `Prop`-valued, but the way I was thinking about it is that `pre_envel_group` defines formal expressions for 1-morphisms at a single object and `pre_envel_group_rel'` defines formal expressions for 2-morphisms. Then `pre_envel_group_rel` uses `Prop` to make all compatible 2-morphisms isomorphic, and the quotient to define `envel_group` makes all related 1-morphisms isomorphic.

The definition of `to_envel_group.map` is simple at the level of 1-morphisms: the formal expressions are interpreted within a specified group `G`, where the expressions have a natural meaning.  But to show this is well defined, the 2-morphisms also need to be evaluated: these need to become equality proofs in `G` (which are the corresponding 2-morphisms).  The 2-morphisms `symm`, `trans`, `congr_mul`, and `congr_inv` each are derived from other 2-morphisms, and `to_envel_group.map_aux.well_def` recursively uses the equality proofs from the sub-2-morphisms to compute the total equality proof.

Experimentally, if you change it to be `Prop`-valued, then Lean complains the induction in `well_def` isn't well-founded.  This seems to make sense because, for example, `symm` would be an involution, so you can't know that inducting on an equivalence would make the equivalence simpler.

Reply:
That's a very helpful explanation (at least on the second reading :-), I learnt something! Could we more or less embed that explanation in the comments? It was not at all obvious to me what was going on here, but now I believe it's both justified and pretty neat.

Reply:
I learned a lot about how it worked by trying to explain it!  Pretty much all I had in mind while writing the code was to try to make it mechanically generalizable to other algebraic structures.  In fact, I started with a construction of the universal enveloping group of a monoid, and since this seems to have worked out, I'm more comfortable creating a PR for that now.

I've added section documentation with an attempt at an even clearer explanation.  It could still be better, but I think I'll need to understand the construction better myself first.